packages/contracts/src/chugsplash/storage.ts

```diff
@@ -0,0 +1,330 @@
+/* External Imports */
+import { fromHexString, remove0x } from '@eth-optimism/core-utils'
+import { BigNumber, ethers } from 'ethers'
+
+// Represents the JSON objects outputted by the Solidity compiler that describe the structure of
+// state within the contract. See
+// https://docs.soliditylang.org/en/v0.8.3/internals/layout_in_storage.html for more information.
+interface SolidityStorageObj {
+  astId: number
+  contract: string
+  label: string
+  offset: number
+  slot: number
+  type: string
+}
+
+// Represents the JSON objects outputted by the Solidity compiler that describe the types used for
+// the various pieces of state in the contract. See
+// https://docs.soliditylang.org/en/v0.8.3/internals/layout_in_storage.html for more information.
+interface SolidityStorageType {
+  encoding: 'inplace' | 'mapping' | 'dynamic_array' | 'bytes'
+  label: string
+  numberOfBytes: number
+  key?: string
+  value?: string
+  base?: string
+  members?: SolidityStorageObj[]
+}
+
+// Container object returned by the Solidity compiler. See
+// https://docs.soliditylang.org/en/v0.8.3/internals/layout_in_storage.html for more information.
+export interface SolidityStorageLayout {
+  storage: SolidityStorageObj[]
+  types: {
+    [name: string]: SolidityStorageType
+  }
+}
+
+interface StorageSlotPair {
+  key: string
+  val: string
+}
+
+/**
+ * Takes a slot value (in hex), left-pads it with zeros, and displaces it by a given offset.
+ * @param val Hex string value to pad.
+ * @param offset Number of bytes to offset from the right.
+ * @return Padded hex string.
+ */
+const padHexSlotValue = (val: string, offset: number): string => {
+  return (
+    '0x' +
+    remove0x(val)
+      .padStart(64 - offset * 2, '0') // Pad the start with 64 - offset zero bytes.
+      .padEnd(64, '0') // Pad the end (up to 64 bytes) with zero bytes.
+      .toLowerCase() // Making this lower case makes assertions more consistent later.
+  )
+}
+
+/**
+ * Retrieves the storageLayout portion of the compiler artifact for a given contract by name. This
+ * function is hardhat specific.
+ * @param hre HardhatRuntimeEnvironment, required for the readArtifactSync function.
+ * @param name Name of the contract to retrieve the storage layout for.
+ * @return Storage layout object from the compiler output.
+ */
+export const getStorageLayout = async (
```

Comment:
It could be good to verify the version of Solidity here / to explicitly specify supported Solidity versions. It would be good to catch an error here where we accidentally upgrade to a new version of Solidity & that breaks the storage slots, as opposed to in post-deploy checks or in prod.

Reply:
Done